packages/vite/src/node/plugins/esbuild.ts

```diff
@@ -243,20 +258,55 @@ export const buildEsbuildPlugin = (config: ResolvedConfig): Plugin => {
         return null
       }
 
-      const res = await transformWithEsbuild(code, chunk.fileName, {
+      const isEsLibBuild = config.build.lib && opts.format === 'es'
+      let options: TransformOptions = {
         ...config.esbuild,
         target: target || undefined,
-        ...(minify
-          ? {
-              // Do not minify ES lib output since that would remove pure annotations
-              // and break tree-shaking
-              // https://github.com/vuejs/core/issues/2860#issuecomment-926882793
-              minify: !(config.build.lib && opts.format === 'es'),
-              treeShaking: true,
-              format: rollupToEsbuildFormatMap[opts.format]
-            }
-          : undefined)
-      })
+        format: rollupToEsbuildFormatMap[opts.format]
+      }
+
+      if (minify) {
+        // If user enable fine-grain minify options, minify with their options instead
+        if (
+          options.minifyIdentifiers === true ||
+          options.minifySyntax === true ||
+          options.minifyWhitespace === true
+        ) {
+          options = { ...options, treeShaking: true }
+          // Do not minify whitespace for ES lib output since that would remove
+          // pure annotations and break tree-shaking
+          // https://github.com/vuejs/core/issues/2860#issuecomment-926882793
+          if (isEsLibBuild) {
+            options = { ...options, minifyWhitespace: true }
+          }
+        } else if (isEsLibBuild) {
+          // Enable all minify except whitespace, which doesn't work
+          options = {
+            ...options,
+            minify: false,
+            minifyIdentifiers: true,
+            minifySyntax: true,
+            treeShaking: true
+          }
+        } else {
+          options = {
+            ...options,
+            minify: true,
+            treeShaking: true
+          }
+        }
+      } else {
+        options = {
+          ...options,
+          minify: false,
+          minifyIdentifiers: false,
+          minifySyntax: false,
+          minifyWhitespace: false,
+          treeShaking: false
+        }
+      }
```

Comment:
It might be better to extract this minify option resolver to a function and add a test for it.

Reply:
Done in 69d81f3